Sources/SourceKitLSP/Swift/SwiftLanguageService.swift

```diff
@@ -968,6 +968,48 @@ extension SwiftLanguageService {
 
     return nil
   }
+
+  public func getReferenceDocument(_ req: GetReferenceDocumentRequest) async throws -> GetReferenceDocumentResponse {
+    switch req.uri.scheme {
+    case "swift-macro-expansion":
+      guard let givenURL = URL(string: req.uri.stringValue) else {
+        throw ResponseError.unknown("Unable to parse given URL")
+      }
+
+      let filePath = "file:/" + givenURL.pathComponents.dropLast().joined(separator: "/")
```

Comment:
Now that I read this, I think it would be good to centralize the creation and parsing of this URL scheme. I’d suggest to create a `MacroExpansionURLScheme` type that has members for all the parameters encoded in the URL scheme and that can be initialized from + converted to a URL. That way it’s easier to make sure that the URL scheme matches in the function that creates it and the functions that interpret it.